botocore/loaders.py

```diff
@@ -1,425 +1,381 @@
-        """
-        Either loads a regular data file (format-specific to subclass) or
-        returns previously loaded data from the cache.
+    The main method used here is ``load_service_model``, which is a
+    convenience method over ``load_data`` and ``determine_latest_version``.
 
-        Returns a dictionary of data from the file.
+    """
+    FILE_LOADER_CLASS = JSONFileLoader
+    # The included models in botocore/data/ that we ship with botocore.
+    BUILTIN_DATA_PATH = os.path.join(BOTOCORE_ROOT, 'data')
+    # For convenience we automatically add ~/.aws/models to the data path.
+    CUSTOMER_DATA_PATH = os.path.join(os.path.expanduser('~'),
+                                      '.aws', 'models')
+
+    def __init__(self, extra_search_paths=None, file_loader=None,
+                 cache=None, include_default_search_paths=True):
+        self._cache = {}
+        if file_loader is None:
+            file_loader = self.FILE_LOADER_CLASS()
+        self.file_loader = file_loader
+        if include_default_search_paths:
+            self._search_paths = [self.CUSTOMER_DATA_PATH,
+                                  self.BUILTIN_DATA_PATH]
+        else:
+            self._search_paths = []
+        if extra_search_paths is not None:
+            self._search_paths.extend(extra_search_paths)
 
-        Usage::
+    @property
+    def search_paths(self):
+        return self._search_paths
 
-            >>> loader = Loader('/path/to/botocore/data')
-            >>> loader.load_data('aws/ec2/2013-02-01')
-            {
-                # ...EC2 service data...
-            }
-            >>> loader.load_data('_endpoints')
-            {
-                # ...Endpoint data...
-            }
+    @instance_cache
+    def list_available_services(self, type_name):
+        """List all known services.
 
-        """
-        # Here, we'll cache it.
-        return self._load_data(data_path)
-
-    def _load_data(self, data_path):
-        # This is the uncached version for use with ``load_service_model``.
-        # We go in-order, returning the first matching path we find
-        # based on the search paths.
-        for possible_path in self.get_search_paths():
-            full_path = os.path.join(
-                possible_path,
-                data_path + self.extension
-            )
-
-            try:
-                return self.file_loader.load_file(full_path)
-            except IOError:
-                continue
+        This will traverse the search path and look for all known
+        services.
 
-        # We didn't find anything that matched on any path.
-        raise DataNotFoundError(data_path=data_path)
+        :type type_name: str
+        :param type_name: The type of the service (service-2,
+            paginators-1, waiters-2, etc).  This is needed because
+            the latest API version available can depend on the service
+            type.  For example, the latest API version available for
```

Comment:
Docstring is a little weird here. Whether it is the latest api version should not matter. It is based on if the model type exists in any of the api versions.